File: interpolation.py
#  ================================================================
#  Created by Gregory Kramida on 9/17/18.
#  Copyright (c) 2018 Gregory Kramida
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#  http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
#  ================================================================

import numpy as np
import math
from utils.point import Point
from utils.sampling import sample_at, focus_coordinates_match, is_outside_narrow_band, sample_flag_at, sample_at_replacement
from utils.printing import BOLD_YELLOW, BOLD_GREEN, RESET


def interpolate_warped_live(canonical_field, warped_live_field, warp_field, gradient_field, band_union_only=False,
                            known_values_only=False,
                            data_gradient_field=None, smoothing_gradient_field=None, substitute_original=False):
    field_size = warp_field.shape[0]
    new_warped_live_field = np.ones_like(warped_live_field)
    for y in range(field_size):
        for x in range(field_size):
            original_live_sdf = warped_live_field[y, x]
            if band_union_only:
                canonical_sdf = canonical_field[y, x]
                if is_outside_narrow_band(original_live_sdf) and is_outside_narrow_band(canonical_sdf):
                    continue
            if known_values_only:
                if original_live_sdf == 1.0:
                    continue

            warped_location = Point(x, y) + Point(coordinates=warp_field[y, x])
            base_point = Point(math.floor(warped_location.x), math.floor(warped_location.y))
            ratios = warped_location - base_point
            inverse_ratios = Point(1.0, 1.0) - ratios

            if substitute_original:
                value00 = sample_at_replacement(warped_live_field, original_live_sdf, point=base_point)
                value01 = sample_at_replacement(warped_live_field, original_live_sdf, point=base_point + Point(0, 1))
                value10 = sample_at_replacement(warped_live_field, original_live_sdf, point=base_point + Point(1, 0))
                value11 = sample_at_replacement(warped_live_field, original_live_sdf, point=base_point + Point(1, 1))
            else:
                value00 = sample_at(warped_live_field, point=base_point)
                value01 = sample_at(warped_live_field, point=base_point + Point(0, 1))
                value10 = sample_at(warped_live_field, point=base_point + Point(1, 0))
                value11 = sample_at(warped_live_field, point=base_point + Point(1, 1))

            interpolated_value0 = value00 * inverse_ratios.y + value01 * ratios.y
            interpolated_value1 = value10 * inverse_ratios.y + value11 * ratios.y
            new_value = interpolated_value0 * inverse_ratios.x + interpolated_value1 * ratios.x
            if 1.0 - abs(new_value) < 1e-6:
                new_value = np.sign(new_value)
                warp_field[y, x] = 0.0
                gradient_field[y, x] = 0.0
                if data_gradient_field is not None:
                    data_gradient_field[y, x] = 0.0
                if smoothing_gradient_field is not None:
                    smoothing_gradient_field[y, x] = 0.0

            if focus_coordinates_match(x, y):
                print("[Interpolation data] ", BOLD_YELLOW,
                      "{:+03.3f}*{:03.3f}, {:+03.3f}*{:03.3f}".format(value00, inverse_ratios.y * inverse_ratios.x,
                                                                      value10, inverse_ratios.y * ratios.x, ),
                      RESET, " original value: ", BOLD_GREEN,
                      original_live_sdf, RESET, sep='')
                print("                     ", BOLD_YELLOW,
                      "{:+03.3f}*{:03.3f}, {:+03.3f}*{:03.3f}".format(value01, ratios.y * inverse_ratios.x,
                                                                      value11, ratios.y * ratios.x),
                      RESET, " final value: ", BOLD_GREEN,
                      new_value, RESET, sep='')
            new_warped_live_field[y, x] = new_value
    np.copyto(warped_live_field, new_warped_live_field)


def interpolate_warped_live_with_flag_info(warped_live_field, warp_field, update_field, flag_field):
    field_size = warp_field.shape[0]
    new_warped_live_field = np.ones_like(warped_live_field)
    for y in range(field_size):
        for x in range(field_size):
            warped_location = Point(x, y) + Point(coordinates=warp_field[y, x])
            base_point = Point(math.floor(warped_location.x), math.floor(warped_location.y))
            ratios = warped_location - base_point
            inverse_ratios = Point(1.0, 1.0) - ratios
            original_value = warped_live_field[y, x]
            value00 = sample_at(warped_live_field, point=base_point)
            flag00 = sample_flag_at(flag_field, point=base_point)
            used_replacement = False
            if flag00 == 0:
                value00 = original_value
                used_replacement = True
            value01 = sample_at(warped_live_field, point=base_point + Point(0, 1))
            flag01 = sample_flag_at(flag_field, point=base_point + Point(0, 1))
            if flag01 == 0:
                value01 = original_value
                used_replacement = True
            value10 = sample_at(warped_live_field, point=base_point + Point(1, 0))
            flag10 = sample_flag_at(flag_field, point=base_point + Point(1, 0))
            if flag10 == 0:
                value10 = original_value
                used_replacement = True
            value11 = sample_at(warped_live_field, point=base_point + Point(1, 1))
            flag11 = sample_flag_at(flag_field, point=base_point + Point(1, 1))
            if flag11 == 0:
                value11 = original_value
                used_replacement = True

            interpolated_value0 = value00 * inverse_ratios.y + value01 * ratios.y
            interpolated_value1 = value10 * inverse_ratios.y + value11 * ratios.y
            interpolated_value = interpolated_value0 * inverse_ratios.x + interpolated_value1 * ratios.x
            if 1.0 - abs(interpolated_value) < 1e-3:
                # if 1.0 - abs(interpolated_value) < 0.05:
                interpolated_value = np.sign(interpolated_value)
                warp_field[y, x] = 0.0
                update_field[y, x] = 0.0

            if focus_coordinates_match(x, y):
                print("[Interpolation data] ", BOLD_YELLOW,
                      "{:+03.3f}*{:03.3f}, {:+03.3f}*{:03.3f}".format(value00, inverse_ratios.y * inverse_ratios.x,
                                                                      value10, inverse_ratios.y * ratios.x, ),
                      RESET, sep='')
                print("                     ", BOLD_YELLOW,
                      "{:+03.3f}*{:03.3f}, {:+03.3f}*{:03.3f}".format(value01, ratios.y * inverse_ratios.x,
                                                                      value11, ratios.y * ratios.x),
                      RESET, " used replacement:", BOLD_GREEN, used_replacement, RESET, " final value: ", BOLD_GREEN,
                      interpolated_value, RESET, sep='')
            new_warped_live_field[y, x] = interpolated_value
    np.copyto(warped_live_field, new_warped_live_field)

File: cpp/src/nonrigid_optimization/interpolation.cpp
//  ================================================================
//  Created by Gregory Kramida on 10/11/18.
//  Copyright (c) 2018 Gregory Kramida
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at

//  http://www.apache.org/licenses/LICENSE-2.0

//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
//  ================================================================
//stdlib
#include <cstdlib>
#include <iostream>

//local
#include "interpolation.hpp"
#include "boolean_operations.hpp"



namespace nonrigid_optimization {

inline float sample_tsdf_value_at(const eig::MatrixXf& tsdf_field, int x, int y) {
	if (x < 0 || x >= tsdf_field.cols() || y < 0 || y >= tsdf_field.rows()) {
		return 1.0;
	}
	return tsdf_field(y, x);
}

inline float sample_tsdf_value_replacing_when_out_of_bounds(const eig::MatrixXf& tsdf_field, int x, int y,
                                                            float replacement_value) {
	if (x < 0 || x >= tsdf_field.cols() || y < 0 || y >= tsdf_field.rows()) {
		return replacement_value;
	}
	return tsdf_field(y, x);
}

eig::MatrixXf interpolate(const eig::MatrixXf& warped_live_field, const eig::MatrixXf& canonical_field,
                          eig::MatrixXf& warp_field_u, eig::MatrixXf& warp_field_v,
                          bool band_union_only, bool known_values_only,
                          bool substitute_original, float truncation_float_threshold) {
	int matrix_size = static_cast<int>(warped_live_field.size());
	const int column_count = static_cast<int>(warped_live_field.cols());
	const int row_count = static_cast<int>(warped_live_field.rows());

	eig::MatrixXf new_live_field(row_count, column_count);

#pragma omp parallel for
	for (int i_element = 0; i_element < matrix_size; i_element++) {
		// Any MatrixXf in Eigen is column-major
		// i_element = x * column_count + y
		div_t division_result = div(i_element, column_count);
		int y = division_result.rem;
		int x = division_result.quot;
		float live_tsdf_value = warped_live_field(i_element);
		if (band_union_only) {
			float canonical_tsdf_value = canonical_field(i_element);
			if (is_outside_narrow_band(live_tsdf_value, canonical_tsdf_value)) {
				new_live_field(i_element) = live_tsdf_value;
				continue;
			}
		}
		if (known_values_only) {
			//TODO assumes 1.0 is the default value
			if (std::abs(live_tsdf_value) == 1.0) {
				new_live_field(i_element) = live_tsdf_value;
				continue;
			}
		}
		float lookup_x = x + warp_field_u(i_element);
		float lookup_y = y + warp_field_v(i_element);
		int base_x = static_cast<int>(std::floor(lookup_x));
		int base_y = static_cast<int>(std::floor(lookup_y));
		float ratio_x = lookup_x - base_x;
		float ratio_y = lookup_y - base_y;
		float inverse_ratio_x = 1.0F - ratio_x;
		float inverse_ratio_y = 1.0F - ratio_y;

		float value00, value01, value10, value11;
		if (substitute_original) {
			value00 = sample_tsdf_value_replacing_when_out_of_bounds(warped_live_field, base_x, base_y,
			                                                         live_tsdf_value);
			value01 = sample_tsdf_value_replacing_when_out_of_bounds(warped_live_field, base_x, base_y + 1,
			                                                         live_tsdf_value);
			value10 = sample_tsdf_value_replacing_when_out_of_bounds(warped_live_field, base_x + 1, base_y,
			                                                         live_tsdf_value);
			value11 = sample_tsdf_value_replacing_when_out_of_bounds(warped_live_field, base_x + 1, base_y + 1,
			                                                         live_tsdf_value);
		} else {
			value00 = sample_tsdf_value_at(warped_live_field, base_x, base_y);
			value01 = sample_tsdf_value_at(warped_live_field, base_x, base_y + 1);
			value10 = sample_tsdf_value_at(warped_live_field, base_x + 1, base_y);
			value11 = sample_tsdf_value_at(warped_live_field, base_x + 1, base_y + 1);
		}

		float interpolated_value0 = value00 * inverse_ratio_y + value01 * ratio_y;
		float interpolated_value1 = value10 * inverse_ratio_y + value11 * ratio_y;
		float new_value = interpolated_value0 * inverse_ratio_x + interpolated_value1 * ratio_x;
		if (1.0 - std::abs(new_value) < truncation_float_threshold) {
			new_value = std::copysign(1.0f, new_value);
			warp_field_u(i_element) = 0.0f; // probably won't work in python due to conversions, test
			warp_field_v(i_element) = 0.0f;
		}

		new_live_field(y, x) = new_value;
	}

	return new_live_field;
}

bp::object
py_interpolate(const eig::MatrixXf& warped_live_field, const eig::MatrixXf& canonical_field, eig::MatrixXf warp_field_u,
               eig::MatrixXf warp_field_v, bool band_union_only, bool known_values_only, bool substitute_original,
               float truncation_float_threshold) {

	eig::MatrixXf new_warped_live_field = interpolate(warped_live_field, canonical_field, warp_field_u, warp_field_v,
	                                                  band_union_only, known_values_only, substitute_original,
	                                                  truncation_float_threshold);

	bp::object warp_field_u_out(warp_field_u);
	bp::object warp_field_v_out(warp_field_v);
	bp::object warped_live_field_out(new_warped_live_field);

	return bp::make_tuple(warped_live_field_out, bp::make_tuple(warp_field_u_out, warp_field_v_out));
}

}//namespace nonrigid_optimization